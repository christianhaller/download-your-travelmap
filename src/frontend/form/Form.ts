import { validate } from "./validate";
import { UrlInput } from "./UrlInput";
import { SubmitButton } from "./SubmitButton";
import type { Response } from "../../backend/interace";
import { Failure } from "../response/failure/Failure";

export class Form {
  private submitButton: SubmitButton;
  private urlInput: UrlInput;
  private el: HTMLFormElement;
  private doc: Document;
  private failure: Failure;

  constructor(document: Document) {
    this.doc = document;
  }

  private async submit(e: Event) {
    e.preventDefault();

    this.doc.dispatchEvent(new CustomEvent("success.hide"));
    this.failure.hide();

    try {
      const url = new URL(this.urlInput.getValue());

      validate(url);
      this.submitButton.loadingStart();

      const [res] = await Promise.all([
        fetch(`/api?url=${url}`),
        import("../success"),
      ]);

      this.submitButton.loadingStop();

      if (res.ok) {
        const detail: Response = await res.json();

        this.doc.dispatchEvent(
          <Event>new CustomEvent("success.show", { detail })
        );

        const newURL = new URL(window.location.href);
        newURL.searchParams.set("url", url.href);

        if (window.location.href !== newURL.href) {
          window.history.pushState(null, "", newURL.href);
        }
      } else {
        throw new Error(
          "profile not found. try something like https://www.tripadvisor.com/Profile/theplanetd"
        );
      }
    } catch (e) {
      this.urlInput.setFocus();
      this.failure.show(e.message);
    }
  }

  public init() {
    this.el = this.doc.querySelector("form");
    this.el.addEventListener("submit", (e: Event) => this.submit(e));
    this.doc.addEventListener(
      "invalid",
      (() => {
        return (e: Event) => {
          e.preventDefault();
          this.urlInput.setFocus();
        };
      })(),
      true
    );

    this.urlInput = new UrlInput().init(this.el);
    this.submitButton = new SubmitButton().init(this.el);
    this.failure = new Failure(this.doc).init();
    return this;
  }
  public autoSubmit(url: string): void {
    this.urlInput.setValue(url);
    this.el.dispatchEvent(new Event("submit", { cancelable: true }));
  }
}
