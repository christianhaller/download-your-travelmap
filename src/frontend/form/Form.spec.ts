import { Form } from "./Form";

let url;
let failureShowMock;

jest.mock("../response/failure/Failure", () => ({
  Failure: jest.fn(() => ({
    init: jest.fn(() => ({
      hide: jest.fn(),
      show: failureShowMock,
    })),
  })),
}));

jest.mock("../success", () => {});

jest.mock("./UrlInput", () => ({
  UrlInput: jest.fn(() => ({
    init: () => ({
      setFocus: () => {},
      getValue: (): string => {
        return url;
      },
      setValue: () => {},
    }),
  })),
}));

jest.mock("./SubmitButton", () => ({
  SubmitButton: jest.fn(() => ({
    init: () => ({
      loadingStart: () => {},
      loadingStop: () => {},
    }),
  })),
}));

const mockSuccessResponse = {
  language: "en",
  username: "christianhaller",
  places: [
    {
      lat: 50.444885,
      lng: 30.536667,
      flags: ["been"],
      city: "Kyiv(Kiev)",
      country: "Ukraine",
    },
    {
      lat: -34.45451,
      lng: -57.83755,
      flags: ["been"],
      city: "Colonia del Sacramento",
      country: "Uruguay",
    },
  ],
};

describe("Form", () => {
  test("submit valid url", async () => {
    document.body.innerHTML =
      "<form><input id='url'/><button type='submit'></button></form>";

    document.addEventListener("success.show", function (opt: any): void {
      expect(opt.detail).toMatchSnapshot();
    });

    const mockJsonPromise = Promise.resolve(mockSuccessResponse);
    const mockFetchPromise = Promise.resolve({
      ok: true,
      json: () => mockJsonPromise,
    });
    global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);
    url = "https://tripadvisor.com/abc";
    new Form(document).init();

    document.querySelector("form").dispatchEvent(new Event("submit"));
  });

  test("submit invalid url", () => {
    document.body.innerHTML =
      "<form><input id='url'/><button type='submit'></button></form>";

    url = "garbage";

    failureShowMock = jest.fn(() => {
      expect(failureShowMock).toHaveBeenCalledWith("Invalid URL: garbage");
    });
    new Form(document).init();

    document
      .querySelector("button")
      .dispatchEvent(new Event("submit", { bubbles: true }));
  });

  test("autosubmit", () => {
    url = "https://tripadvisor.com/abc";
    const mockJsonPromise = Promise.resolve(mockSuccessResponse);
    const mockFetchPromise = Promise.resolve({
      ok: true,
      json: () => mockJsonPromise,
    });
    global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);
    document.body.innerHTML =
      "<form><input id='url'/><button type='submit'></button></form>";
    document.addEventListener("success.show", function (opt: any): void {
      expect(opt.detail).toMatchSnapshot();
    });
    new Form(document)
      .init()
      .autoSubmit("https://www.tripadvisor.com/members/abc");
  });

  test("fetch failed", () => {
    const mockJsonPromise = Promise.resolve(mockSuccessResponse);
    const mockFetchPromise = Promise.resolve({
      ok: false,
      json: () => mockJsonPromise,
    });

    failureShowMock = jest.fn(() => {
      expect(failureShowMock).toHaveBeenCalled();
    });
    global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);
    document.body.innerHTML =
      "<form><input id='url'/><button type='submit'></button></form>";

    url = "https://www.tripadvisor.it/members/me";
    new Form(document).init();

    document
      .querySelector("button")
      .dispatchEvent(new Event("submit", { bubbles: true }));
  });

  test("invalid by browser", () => {
    document.body.innerHTML =
      "<form><input id='url'/><button type='submit'></button></form>";
    new Form(document).init();
    document
      .querySelector("form")
      .dispatchEvent(new Event("invalid", { bubbles: true }));

    expect(document.activeElement === document.querySelector("#url"));
  });
});
