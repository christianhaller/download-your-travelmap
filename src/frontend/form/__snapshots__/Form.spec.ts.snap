// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form autosubmit 1`] = `
{
  "language": "en",
  "places": [
    {
      "city": "Kyiv(Kiev)",
      "country": "Ukraine",
      "flags": [
        "been",
      ],
      "lat": 50.444885,
      "lng": 30.536667,
    },
    {
      "city": "Colonia del Sacramento",
      "country": "Uruguay",
      "flags": [
        "been",
      ],
      "lat": -34.45451,
      "lng": -57.83755,
    },
  ],
  "username": "christianhaller",
}
`;

exports[`Form autosubmit 2`] = `
{
  "language": "en",
  "places": [
    {
      "city": "Kyiv(Kiev)",
      "country": "Ukraine",
      "flags": [
        "been",
      ],
      "lat": 50.444885,
      "lng": 30.536667,
    },
    {
      "city": "Colonia del Sacramento",
      "country": "Uruguay",
      "flags": [
        "been",
      ],
      "lat": -34.45451,
      "lng": -57.83755,
    },
  ],
  "username": "christianhaller",
}
`;

exports[`Form submit valid url 1`] = `
{
  "language": "en",
  "places": [
    {
      "city": "Kyiv(Kiev)",
      "country": "Ukraine",
      "flags": [
        "been",
      ],
      "lat": 50.444885,
      "lng": 30.536667,
    },
    {
      "city": "Colonia del Sacramento",
      "country": "Uruguay",
      "flags": [
        "been",
      ],
      "lat": -34.45451,
      "lng": -57.83755,
    },
  ],
  "username": "christianhaller",
}
`;
